auto store remote to local pool

lisiG9 opened the pull request:
feat: auto-put remote cache to LocalCPUBackend

What this PR does / why we need it:
In the PD-separated with lmcache + Mooncake, data fetched by the D-side from remote sources (e.g., MoonCache) is not automatically stored in the local CPU pool (LocalCPUBackend). This causes repeated remote downloads of the same prefix data during multi-turn dialogues, leading to high latency.

Solution:
Only modify the batched_get function in storage_manager.py
Auto-write fetched data to the local CPU cache when retrieved from remote backends (non-LocalCPUBackend/PDBackend)
Avoid duplicate storage: Leverage existing local cache existence checks, no redundant logic added

Compatibility:
Tested and verified on NVIDIA GPU + Ascend NPU

If applicable:

- [ ] this PR contains user facing changes: No (no new configurations/interfaces added, only underlying logic optimization)
- [ ] this PR contains unit tests: No (manual validation completed; can supplement per project guidelines)

lisiG9 commented:
@maobaolong  @ApostaC @sammshen Could you please check if this MR is correct?

ApostaC commented:
Hey @lisiG9 , thanks for the contribution! IIUC, the current storage manager already has very similar functionality. Basically, when we are using LocalCPUBackend with RemoteBackend/LocalDiskBackend, the data loaded from lower-tier memory will be put into LocalCPUBackend.

lisiG9 commented:
Hi @ApostaC , when I was using mooncake for cross-node cache management with large model PD (Prefiller-Decoder) separation (with both `local_cpu` and `remote_url` registered in the configuration file), I found that only the **prefiller side** would store cache in both `localcpu_backend` and `remote_backend`, while the **decoder side** always fetches data from `remote_backend`.

Here is part of my shared configuration:
```bash
COMMON_ARGS=(
    "--trust_remote_code"
    "--max-num-seqs=192"
    "--max_model_len=20480"
    "--block-size=32"
    "--gpu-memory-utilization=0.9"
    "--disable-log-requests"
    "--no-enable-prefix-caching"
    "--enforce_eager"
)

CUDA_VISIBLE_DEVICES=$DEVICE_ID python3 -m vllm.entrypoints.openai.api_server \
    --model "$MODEL_PATH" \
    --port "$PORT" \
    --tensor_parallel_size "$TENSOR_PARALLEL_SIZE" \
    "${COMMON_ARGS[@]}" \
    --kv-transfer-config \
"{\"kv_connector\":\"LMCacheConnectorV1\",\"kv_role\":\"$KV_ROLE_VALUE\",\"kv_connector_extra_config\": {\"discard_partial_chunks\": false, \"lmcache_rpc_port\": \"$LMCACHE_RPC_PORT_NAME\"}}"
```

And the **LMCache configuration file**:
```yaml
chunk_size: 256
remote_url: "mooncakestore://localhost:50011/"
remote_serde: "naive"
local_cpu: True
max_local_cpu_size: 10
use_layerwise: False
save_decode_cache: False
enable_async_loading: False
numa_mode: "auto"
auto_store_remote_to_local: False

extra_config:
  local_hostname: "localhost"
  protocol: "tcp"
  device_name: ""
  metadata_server: "http://localhost:8082/metadata"
  master_server_address: "localhost:50011"
  global_segment_size: 53687091200 # 50G
  local_buffer_size: 0    # rely on LMCache local_cpu as the buffer
  transfer_timeout: 60
  save_chunk_meta: False
  use_exists_sync: True
```

In the `_process_tokens_internal` function of `cache_engine.py`, after starting the loop with `for location, blocks in block_mapping.items():`, I added logging to print relevant information:  
`logger.info(f"Retrieved {len(memory_objs)} memory objects from {location}")`

The logs clearly demonstrate the issue and the effect of my improvement:

### Prefiller (P-side) Logs (fetches duplicate cache from LocalCPUBackend)
```log
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:29,827] LMCache INFO: Reqid: chatcmpl-8e1011b0d1d69ad5, Total tokens 8210, LMCache hit tokens: 0, need to load: 0 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:30,151] LMCache INFO: Storing KV cache for 2048 out of 2048 tokens (skip_leading_tokens=0) for request chatcmpl-8e1011b0d1d69ad5 (vllm_v1_adapter.py:1295:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:30,162] LMCache INFO: Stored 2048 out of total 2048 tokens. size: 0.1094 gb, cost 10.7586 ms, throughput: 10.1662 GB/s; offload_time: 10.6520 ms, put_time: 0.1067 ms (cache_engine.py:356:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:30,515] LMCache INFO: Storing KV cache for 2048 out of 4096 tokens (skip_leading_tokens=2048) for request chatcmpl-8e1011b0d1d69ad5 (vllm_v1_adapter.py:1295:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:30,527] LMCache INFO: Stored 2048 out of total 2048 tokens. size: 0.1094 gb, cost 11.9230 ms, throughput: 9.1734 GB/s; offload_time: 11.7349 ms, put_time: 0.1881 ms (cache_engine.py:356:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:30,911] LMCache INFO: Storing KV cache for 2048 out of 6144 tokens (skip_leading_tokens=4096) for request chatcmpl-8e1011b0d1d69ad5 (vllm_v1_adapter.py:1295:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:30,923] LMCache INFO: Stored 2048 out of total 2048 tokens. size: 0.1094 gb, cost 11.7943 ms, throughput: 9.2735 GB/s; offload_time: 11.6749 ms, put_time: 0.1195 ms (cache_engine.py:356:lmcache.v1.cache_engine)
(APIServer pid=44717) INFO 12-10 02:06:31 [loggers.py:236] Engine 000: Avg prompt throughput: 821.1 tokens/s, Avg generation throughput: 0.1 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 5.7%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 0.0%
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:31,318] LMCache INFO: Storing KV cache for 2048 out of 8192 tokens (skip_leading_tokens=6144) for request chatcmpl-8e1011b0d1d69ad5 (vllm_v1_adapter.py:1295:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:31,331] LMCache INFO: Stored 2048 out of total 2048 tokens. size: 0.1094 gb, cost 12.2029 ms, throughput: 8.9630 GB/s; offload_time: 12.0175 ms, put_time: 0.1854 ms (cache_engine.py:356:lmcache.v1.cache_engine)
(APIServer pid=44717) INFO:     127.0.0.1:43332 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:38,423] LMCache INFO: Reqid: chatcmpl-8463dcda3c1cbc1e, Total tokens 8210, LMCache hit tokens: 8192, need to load: 8192 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:38,428] LMCache INFO: Retrieved 32 memory objects from LocalCPUBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:38,467] LMCache INFO: Retrieved 8192 out of 8192 required tokens (from 8192 total tokens). size: 0.4375 gb, cost 40.3594 ms, throughput: 10.8401 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:38,492] LMCache INFO: Storing KV cache for 18 out of 8210 tokens (skip_leading_tokens=8192) for request chatcmpl-8463dcda3c1cbc1e (vllm_v1_adapter.py:1295:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:38,493] LMCache INFO: Stored 18 out of total 18 tokens. size: 0.0010 gb, cost 0.6299 ms, throughput: 1.5261 GB/s; offload_time: 0.5465 ms, put_time: 0.0834 ms (cache_engine.py:356:lmcache.v1.cache_engine)
(APIServer pid=44717) INFO 12-10 02:06:41 [loggers.py:236] Engine 000: Avg prompt throughput: 1641.9 tokens/s, Avg generation throughput: 0.2 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 33.3%
(APIServer pid=44717) INFO:     127.0.0.1:43332 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:43,237] LMCache INFO: Reqid: chatcmpl-804422073803ce7c, Total tokens 8210, LMCache hit tokens: 8210, need to load: 8209 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:43,241] LMCache INFO: Retrieved 33 memory objects from LocalCPUBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45009) [2025-12-10 02:06:43,281] LMCache INFO: Retrieved 8210 out of 8210 required tokens (from 8210 total tokens). size: 0.4385 gb, cost 39.9655 ms, throughput: 10.9710 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
(APIServer pid=44717) INFO 12-10 02:06:51 [loggers.py:236] Engine 000: Avg prompt throughput: 821.0 tokens/s, Avg generation throughput: 0.1 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 49.9%
(APIServer pid=44717) INFO 12-10 02:07:01 [loggers.py:236] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 49.9%
```

### Decoder (D-side) Logs (before improvement, always fetches from RemoteBackend)
```log
(EngineCore_DP0 pid=45013) [2025-12-10 02:06:38,556] LMCache INFO: Reqid: chatcmpl-96895635e18f40c3, Total tokens 8210, LMCache hit tokens: 8210, need to load: 8209 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45013) [2025-12-10 02:06:38,913] LMCache INFO: Retrieved 33 memory objects from RemoteBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45013) [2025-12-10 02:06:38,953] LMCache INFO: Retrieved 8210 out of 8210 required tokens (from 8210 total tokens). size: 0.4385 gb, cost 391.7658 ms, throughput: 1.1192 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
(APIServer pid=44716) INFO 12-10 02:06:41 [loggers.py:236] Engine 000: Avg prompt throughput: 1641.9 tokens/s, Avg generation throughput: 10.2 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 7.6%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 74.9%
(APIServer pid=44716) INFO:     127.0.0.1:53132 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(EngineCore_DP0 pid=45013) [2025-12-10 02:06:43,376] LMCache INFO: Reqid: chatcmpl-89317fecb197b785, Total tokens 8210, LMCache hit tokens: 8210, need to load: 8209 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=45013) [2025-12-10 02:06:43,728] LMCache INFO: Retrieved 33 memory objects from RemoteBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=45013) [2025-12-10 02:06:43,768] LMCache INFO: Retrieved 8210 out of 8210 required tokens (from 8210 total tokens). size: 0.4385 gb, cost 386.8799 ms, throughput: 1.1333 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
(APIServer pid=44716) INFO 12-10 02:06:51 [loggers.py:236] Engine 000: Avg prompt throughput: 821.0 tokens/s, Avg generation throughput: 9.3 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 81.2%
(APIServer pid=44716) INFO 12-10 02:07:01 [loggers.py:236] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 81.2%
```

### Decoder (D-side) Logs (after my improvement, fetches from local cache pool)
```log
(EngineCore_DP0 pid=43511) [2025-12-10 02:00:36,296] LMCache INFO: Reqid: chatcmpl-b228dd88b47b7553, Total tokens 8210, LMCache hit tokens: 8210, need to load: 8209 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(APIServer pid=43191) INFO 12-10 02:00:36 [loggers.py:236] Engine 000: Avg prompt throughput: 821.0 tokens/s, Avg generation throughput: 6.3 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 49.9%
(EngineCore_DP0 pid=43511) [2025-12-10 02:00:36,647] LMCache INFO: Retrieved 33 memory objects from RemoteBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=43511) [2025-12-10 02:00:36,648] LMCache INFO: [Remote→Local] Stored 33 keys from RemoteBackend to LocalCPUBackend (cache_engine.py:1355:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=43511) [2025-12-10 02:00:36,688] LMCache INFO: Retrieved 8210 out of 8210 required tokens (from 8210 total tokens). size: 0.4385 gb, cost 388.3659 ms, throughput: 1.1290 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
(APIServer pid=43191) INFO 12-10 02:00:46 [loggers.py:236] Engine 000: Avg prompt throughput: 821.0 tokens/s, Avg generation throughput: 6.6 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 66.6%
(APIServer pid=43191) INFO 12-10 02:00:56 [loggers.py:236] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Waiting: 0 reqs, GPU KV cache usage: 0.0%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 66.6%
(APIServer pid=43191) INFO:     127.0.0.1:36776 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:23,854] LMCache INFO: Reqid: chatcmpl-89a8dacddcab3f98, Total tokens 8211, LMCache hit tokens: 8211, need to load: 8210 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:23,861] LMCache INFO: Retrieved 32 memory objects from LocalCPUBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:23,866] LMCache INFO: Retrieved 1 memory objects from RemoteBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:23,866] LMCache INFO: [Remote→Local] Stored 1 keys from RemoteBackend to LocalCPUBackend (cache_engine.py:1355:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:23,905] LMCache INFO: Retrieved 8211 out of 8211 required tokens (from 8211 total tokens). size: 0.4385 gb, cost 44.9764 ms, throughput: 9.7499 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
(APIServer pid=43191) INFO 12-10 02:03:26 [loggers.py:236] Engine 000: Avg prompt throughput: 821.1 tokens/s, Avg generation throughput: 3.9 tokens/s, Running: 1 reqs, Waiting: 0 reqs, GPU KV cache usage: 7.6%, Prefix cache hit rate: 0.0%, External prefix cache hit rate: 74.9%
(APIServer pid=43191) INFO:     127.0.0.1:36776 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:29,043] LMCache INFO: Reqid: chatcmpl-878ee638693b5ca1, Total tokens 8210, LMCache hit tokens: 8210, need to load: 8209 (vllm_v1_adapter.py:1520:lmcache.integration.vllm.vllm_v1_adapter)
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:29,047] LMCache INFO: Retrieved 33 memory objects from LocalCPUBackend (cache_engine.py:1331:lmcache.v1.cache_engine)
(EngineCore_DP0 pid=43511) [2025-12-10 02:03:29,087] LMCache INFO: Retrieved 8210 out of 8210 required tokens (from 8210 total tokens). size: 0.4385 gb, cost 40.3952 ms, throughput: 10.8543 GB/s; (cache_engine.py:586:lmcache.v1.cache_engine)
```

ApostaC commented:
Interesting. @YaoJiayi Can you take a look at this?

ApostaC commented:
Btw, there is an on-going work of p2p KV cache sharing, which enable you to share the KV cache across local_cpu_backend on different nodes. It may be a better alternative than using mooncake + local_cpu_backend.
Timeline-wise, it will be landed before the end of this month.

lisiG9 commented:
Hi @ApostaC @YaoJiayi , thanks for sharing the upcoming p2p KV cache feature—it’ll be a big step forward for cross-node cache in memory-rich homogeneous clusters.

My improvement is a solid complement for specific production scenarios:

1. **Lightweight & low-maintenance**  
It’s just a small Remote→Local sync for decoder cache fetch, no heavy libs like NCCL/P2P RPC needed, so it’s easy to maintain and won’t break the existing mooncake + local_cpu stack.
2. **Modular for high-pressure workloads**  
Teams (including ours) use mooncake and other cache tools for tiered CPU/SSD scheduling under high-concurrency multi-turn loads. This fix fills PD-separated cache gaps and gives a modular hook for custom optimizations without rewriting core logic.

It’s not a rival to p2p—it’s just for teams wanting easy, toolchain-linked inference optimization.

Also, would you happen to have any specific docs or patches to share regarding the p2p KV sharing feature?

ApostaC commented:
Hey @lisiG9 , a few quick comment on this PR:
1. Please fix the linter issues by running pre-commit
2. The storage backends are "implementation details" and should be hidden behind the `StorageManager`. It would be better if the logic can be moved from cache engine to storage manager.

lisiG9 commented:
Hi @ApostaC , thanks for the feedback!

We’ve fixed all linter issues via pre-commit.

We split the logic: StorageManager handles storage backend details (e.g., local/remote pools), while CacheEngine manages cache policy (e.g., remote→local cache moves) – this keeps storage backends hidden behind StorageManager as you suggested.

liwei-rk reviewed (approved):
Please check  get function.

lisiG9 committed:
align batched_get with get by adding local cache logic

Signed-off-by: lisiG9 <2442737518@qq.com>

align batched_get with get

Signed-off-by: lisiG9 <2442737518@qq.com>

pair valid keys with non - None memory objects

Signed-off-by: lisiG9 <2442737518@qq.com>

refactor: extract none obj filter logic into standalone function

Signed-off-by: lisiG9 <2442737518@qq.com>

refactor: extract submit tasks into standalone function

Signed-off-by: lisiG9 <2442737518@qq.com>

Only write back if we found ALL keys

Signed-off-by: liwei <2442737518@qq.com>
Signed-off-by: lisiG9 <2442737518@qq.com>

storage_manager: align batched_get remote-to-local write-back with get() + add comments & type cast

Signed-off-by: lisiG9 <2442737518@qq.com>

change TODO

Signed-off-by: lisiG9 <2442737518@qq.com>

Apply suggestion from @ApostaC

Co-authored-by: Yihua Cheng <yihuac@vllm.ai>
Signed-off-by: lisiG9 <86335240+lisiG9@users.noreply.github.com>
Signed-off-by: lisiG9 <2442737518@qq.com>

Adjust the if condition

Signed-off-by: lisiG9 <2442737518@qq.com>

lisiG9 commented:
Hi @ApostaC @YaoJiayi ,I aligned the batched_get logic fully with get by removing redundant conditional checks.

The local CPU backend write-back logic in batched_get now matches exactly with the single-key get method for consistency.

sammshen commented:
@lisiG9

the async codepath also has a currently unfilled single load callback that is a placeholder for this functionality.

just a suggestion: 

how about if we only perform write-back when all requested keys are found?

sammshen commented:
```python
        for backend_name, storage_backend in self.get_active_storage_backends(location):
            memory_objs = storage_backend.batched_get_blocking(keys)
            if memory_objs:
                if (
                    backend_name not in ["LocalCPUBackend", "PDBackend"]
                    and "LocalCPUBackend" in self.storage_backends
                ):
                    # Only write back if we found ALL keys (no Nones)
                    if None not in memory_objs:
                        logger.debug(
                            f"Storing {len(keys)} objects from"
                            f" {backend_name} to LocalCPUBackend"
                        )
                        self.storage_backends["LocalCPUBackend"].batched_submit_put_task(
                            keys, memory_objs
                        )
                return memory_objs
```

sammshen commented:
the reasoning is that having `None` being retrieved is an error case.

sammshen commented:
```python
    def prefetch_single_done_callback(
        self,
        future: asyncio.Future,
        keys: list[CacheEngineKey],
        backend_name: str,
    ) -> None:
        """
        Callback function when a single prefetch task
        (i.e., prefetching from a single backend) is done.
        """
        # TODO(Jiayi): support write-back policy here
        pass
 ```
 
 this is the other place marked by Jiayi in teh storage manager for WB / cache promotion previously

lisiG9 commented:
Thanks @ApostaC  and @sammshen  for your guidance! I've updated the `batched_get` logic as suggested.

We are continuing to optimize the multi-level cache mechanism. My friend @ZER0921  will follow up on the TODO item for `prefetch_single_done_callback` in the next step.

lisiG9 commented:
The current changes differ significantly from the initial submission. To avoid confusion, I’ve updated the PR description at the top

ZER0921 commented:
> Thanks @ApostaC and @sammshen for your guidance! I've updated the `batched_get` logic as suggested.
> 
> We are continuing to optimize the multi-level cache mechanism. My friend @ZER0921 will follow up on the TODO item for `prefetch_single_done_callback` in the next step.

I will fix this TODO as soon as possible.

sammshen reviewed (approved):
LGTM! Looking forward to the follow-up!

YaoJiayi reviewed (commented):
Thanks! The code here looks pretty hacky. To me it should be part of the caching policy. It's okay to keep it like this for now, but please add some comments on explaining the code and say we'll refactor the code later. Also, this write-back mechanism is tuned on by default but I don't think all users want this.

lisiG9 commented:
Hi @YaoJiayi! Thanks for your feedback 🙏

I've updated the code with detailed comments and adjustments:
1. Added the type cast as required by GitHub's code quality checks.
2. Marked a TODO to refactor this logic into the caching policy module later, as you suggested.

One quick question: Since the single-key `get()` method doesn't have a toggle for this write-back behavior, I'm wondering if we should keep consistency and not add a config switch for the batched version either. What are your thoughts on this, @YaoJiayi and @ApostaC?

Given that putting cache into the localCPU_Backend is zero-cost (similar to prefix-cache) 
the cache gets evicted naturally when the cache is full, and can’t be stored anyway if the local_cpu is disabled
 this also aligns with the default multi-level cache logic and matches user expectations.

ningziwen commented:
This should be able to fix the second issue in https://github.com/LMCache/LMCache/issues/2017.

lisiG9 commented:
Hi @maobaolong, @ApostaC, @YaoJiayi! This change is crucial for us. Following up on the previous feedback, could you please take a look and merge it if appropriate at your earliest convenience?

ApostaC commented:
@lisiG9 Hey, I pinged Jiayi offline, and he will get back to this PR very soon

ApostaC commented on lmcache/v1/storage_backend/storage_manager.py:
```diff
@@ -475,6 +476,28 @@ def batched_get(
         for backend_name, storage_backend in self.get_active_storage_backends(location):
             memory_objs = storage_backend.batched_get_blocking(keys)
             if memory_objs:
```
Actually, I'm wondering why we have an `if memory_objs` check here? What's the difference between returning None and returning an empty list for this function @YaoJiayi @sammshen ?

sammshen replied:
the cache engine is ok to receive either empty list or None placeholders:

```python
   def _process_tokens_internal:
          ...
            # strict = False here
            for (key, start, end), memory_obj in zip(blocks, memory_objs, strict=False):
                 if memory_obj is None: 
                     break
```

sammshen replied:
the `batched_get_non_blocking() -> List[MemoryObj]` does not return None, maybe we could unify this

ApostaC commented on lmcache/v1/storage_backend/storage_manager.py:
```diff
@@ -475,6 +476,28 @@ def batched_get(
+                    and "LocalCPUBackend" in self.storage_backends
+                ):
+                    # Only write back if we found ALL keys (no Nones)
+                    if None not in memory_objs:
```
In practice, this could be pretty slow if `memory_objs` is a long list (it's an O(N)) operation.

One potential hacky solution is to check whether memory_objs[-1] is None. This is based on the assumption that if the returned memory objects should be something like `[obj1, obj2, ..., obj_i, None, None, None, ..., None]`. Not sure if it's true or not, though @YaoJiayi

ApostaC commented on lmcache/v1/storage_backend/storage_manager.py:
```diff
@@ -475,6 +476,28 @@ def batched_get(
+                        logger.debug(
+                            f"Storing {len(keys)} objects from"
+                            f" {backend_name} to LocalCPUBackend"
+                        )
```
```suggestion
                        logger.debug(
                            "Storing %s objects from %s to LocalCPUBackend",
                            len(keys), backend_name
                        )
```

ApostaC commented on lmcache/v1/storage_backend/storage_manager.py:
```diff
@@ -475,6 +476,28 @@ def batched_get(
+                    and "LocalCPUBackend" in self.storage_backends
+                ):
+                    # Only write back if we found ALL keys (no Nones)
+                    if None not in memory_objs:
```
In general, there are too many nested "if" without any "else" branch here. We should consider merging them into a single if condition check.

lisiG9 replied:
Hey @ApostaC, thanks for your suggestion – it’s right to avoid nesting multiple if statements. 
I plan to merge the logic after @YaoJiayi confirms the checking whether there are None values in memory_objs.

ApostaC reviewed (commented):
Thanks @lisiG9 , please see my comment below

liwei-rk commented:
we can make the batched_get_blocking return empty, if None exists.

the current batched_get_blocking is a  very simple loop by calling get_blocking one bye one.

YaoJiayi reviewed (approved):
it's okay to have a hacky implementation like this for now. @ApostaC @sammshen We need to have a module to control the cache policy among different tiers in the future.

lisiG9 commented:
Hi @yaojiayi @ApostaC @sammshen, thanks for your help! I’ve adjusted the nested if logic – could you help re-approve the workflows for this PR?

lisiG9 commented:
Hi @sammshen, Sorry, I clicked rebase just now, and now auto-merge has been disabled. Could you please re-enable it for me?

lisiG9 commented:
> Hi @sammshen, Sorry, I clicked rebase just now, and now auto-merge has been disabled. Could you please re-enable it for me?

Could you help check if the pending buildkite/comprehensive-tests can be triggered? Thanks a lot! @sammshen @YaoJiayi

sammshen commented:
it takes a while to trigger, I can't speed it up

lisiG9 commented:
> it takes a while to trigger, I can't speed it up
Could this be because the full label hasn't been added to the PR yet?

lisiG9 commented:
Hey @sammshen @YaoJiayi  , Three days have passed with no progress. Could you help look into this?

sammshen commented:
Sorry the auto merger got cancelled by rebase. It should work now

sammshen commented:
Sorry the auto merger got cancelled by rebase. It should work now
